Making metrics splits configurable from cConf at creation time

CuriousVini opened the pull request:
Issue: https://issues.cask.co/browse/CDAP-12297
Build: https://builds.cask.co/browse/CDAP-RUT1244

Tested on cluster by updating the dataset properties of metrics table.

CuriousVini committed:
Making metrics splits configurable from cConf at creation time

anew commented on cdap-common/src/main/resources/cdap-default.xml:
```diff
@@ -2097,10 +2097,11 @@
+    <value>16</value>
     <description>
-      Enable or disable scanning v2 metrics tables. By default set to false.
+      Number of splits for all v3 metrics tables. This property can only be changed when v3 tables are created for the
```
I would just drop the v3 from this description. Going forward, this is the metrics table.

anew replied:
Also, users do not know when the tables are created for the first time. Perhaps say: can only be changed before CDAP starts up for the first time and creates the metrics tables.

CuriousVini replied:
fixed

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java:
```diff
@@ -95,11 +96,19 @@ public HBaseMetricsTable(DatasetContext datasetContext, DatasetSpecification spe
     this.scanExecutor = null;
     this.rowKeyDistributor = null;
     if (isV3Table) {
+      // If v3 table is already created, we will ignore splits value from cConf to avoid modifying splits after creation
```
I am wondering how often this warning will be logged? Does metrics service instantiate metrics tables a lot? Possibly for every query f so, then this can flood the logs. It may be a better idea to make this check as part of startup check, or once when metrics processor starts up.

CuriousVini replied:
This warnings will be logged for each thread having this dataset. Even though it does not flood the master logs on cluster, it does get logged more than 4 times from different threads I believe. 

The problem in adding it in `MessagingMetricsProcessorService` is the dataset spec is not exposed by `MetricsTable` dataset. Either we can expose it from `MetricsTable` so that we can use it in metrics processor or we can add a check in `HBaseMetricsTableDefinition` with log sampler?

anew replied:
It's ok to expose a method to get the spec (or only the number of splits). MetricsTable is an internal dataset and only used by CDAP.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java:
```diff
@@ -115,7 +124,7 @@ public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
       };
 
-      int maxScanThread = spec.getIntProperty(Constants.Metrics.METRICS_HBASE_MAX_SCAN_THREADS, 96);
+      int maxScanThread = cConf.getInt(Constants.Metrics.METRICS_HBASE_MAX_SCAN_THREADS, 96);
```
Is this property in cdap-default.xml? In that case you don't need a default here.

anew commented on cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseMetricsTable.java:
```diff
@@ -125,7 +134,8 @@ public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
       this.rowKeyDistributor = new RowKeyDistributorByHashPrefix(
-        new RowKeyDistributorByHashPrefix.OneByteSimpleHash(Constants.Metrics.METRICS_HBASE_SPLITS));
+        new RowKeyDistributorByHashPrefix.
+          OneByteSimpleHash(spec.getIntProperty(Constants.Metrics.METRICS_HBASE_TABLE_SPLITS, 16)));
```
because the dataset is always created with this property - do you need the default here? If you pass in a default, it should be the value from cConf. I think that can only happen when people upgrade from 4.3RC to 4.3GA - their metrics tables would have been created without the property. Oh is that why you use 16 as the default?

CuriousVini replied:
1.) If I use cConf value instead of default 16, and if cConf value is modified from 16 (which was 16 in GA), then we will get wrong distributed keys. Should we keep it as 16 only? 
2.) yea the default will be for RC to GA

anew reviewed (commented):
please see my comments

CuriousVini commented:
Thanks @anew for the review. I have addressed comments and answered questions. Please take another pass when you get a chance. 

Tested the changes on cluster.

anew commented:
Hi @CuriousVini, we talked about checking the splits only in the metrics processor and not in the dataset itself, in order to avoid flooding the logs with warnings when the metrics dataset is instantiated times and times again. Are you going to make that change? 

Other changes look good to me.

CuriousVini commented:
@anew Thanks for the review. I have now moved validation to Metrics processor. Please take another pass when you get a chance, Thanks.

Tested on cluster. 

![image](https://user-images.githubusercontent.com/14131070/29535827-409e361a-8670-11e7-99d8-033dd6012c8f.png)

CuriousVini commented:
Build is green. https://builds.cask.co/browse/CDAP-RUT1244-4

anew commented:
LGTM

CuriousVini commented:
Build is green. Squashing and merging.
